src/datastore/postgres-store.ts

```diff
@@ -2936,6 +3096,306 @@ export class PgDataStore extends (EventEmitter as { new (): DataStoreEventEmitte
     });
   }
 
+  async updateNames(client: ClientBase, bnsName: DbBNSName) {
+    const {
+      name,
+      address,
+      registered_at,
+      expire_block,
+      zonefile_hash,
+      zonefile,
+      namespace_id,
+      latest,
+      tx_id,
+      status,
+      canonical,
+      index_block_hash,
+      atch_resolved,
+    } = bnsName;
+    await client.query(`UPDATE names SET latest = $1 WHERE name= $2`, [false, name]);
+
+    await client.query(
+      `
+        INSERT INTO names(
+      name, address, registered_at, expire_block, zonefile_hash, zonefile, namespace_id, latest, tx_id, status, canonical, index_block_hash, atch_resolved
+        ) values($1, $2, $3, $4, $5, $6, $7, $8,$9, $10, $11, $12, $13)
+        `,
+      [
+        name,
+        address,
+        registered_at,
+        expire_block,
+        zonefile_hash,
+        zonefile,
+        namespace_id,
+        latest,
+        hexToBuffer(tx_id ? tx_id : '0x'),
+        status,
+        canonical,
+        index_block_hash,
+        atch_resolved,
+      ]
+    );
+  }
+
+  async updateNamespaces(client: ClientBase, bnsNamespace: DbBNSNamespace) {
+    const {
+      namespace_id,
+      launched_at,
+      address,
+      reveal_block,
+      ready_block,
+      buckets,
+      base,
+      coeff,
+      nonalpha_discount,
+      no_vowel_discount,
+      lifetime,
+      status,
+      latest,
+      tx_id,
+      canonical,
+      index_block_hash,
+    } = bnsNamespace;
+    await client.query(`UPDATE namespaces SET latest = $1 WHERE namespace_id= $2`, [
+      false,
+      namespace_id,
+    ]);
+
+    await client.query(
+      `
+        INSERT INTO namespaces(
+          namespace_id, launched_at, address, reveal_block, ready_block, buckets,
+          base,coeff,nonalpha_discount,no_vowel_discount,lifetime,status,latest,
+          tx_id, canonical, index_block_hash
+        ) values($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)
+        `,
+      [
+        namespace_id,
+        launched_at,
+        address,
+        reveal_block,
+        ready_block,
+        buckets,
+        base,
+        coeff,
+        nonalpha_discount,
+        no_vowel_discount,
+        lifetime,
+        status,
+        latest,
+        hexToBuffer(tx_id ? tx_id : '0x'),
+        canonical,
+        index_block_hash,
+      ]
+    );
+  }
+
+  async getNamespaceList() {
+    const queryResult = await this.pool.query(
+      `
+      SELECT namespace_id
+      FROM namespaces
+      WHERE 
+      latest = true
+      AND 
+      canonical = true
+      ORDER BY 
+      namespace_id
+      `
+    );
+
+    const results = queryResult.rows.map(r => r.namespace_id);
+    return { results };
+  }
+
+  async getNamespaceNamesList(args: { namespace: string; page: number }) {
+    const offset = args.page * 100;
+    const queryResult = await this.pool.query(
+      `
+      SELECT name
+      FROM names
+      WHERE namespace_id = $1
```

Comment:
Is this function intended to return historical names? If not, this needs a `latest = true`.

Also needs a `canonical = false` to ignore orphaned data.

Reply:
shouldn't this be `canonical = true`?

Reply:
Oops, yes `canonical = true`